src/mc_tasks/ImpedanceTask.cpp

```diff
@@ -122,7 +123,22 @@ void ImpedanceTask::update(mc_solver::QPSolver & solver)
     deltaCompPoseW_.rotation() = aaDeltaCompRot.toRotationMatrix();
   }
 
-  // 4. Set compliance values to the targets of SurfaceTransformTask
+  // 4. Update deltaCompPoseW_ in hold mode
+  if(hold_)
+  {
+    // holdOffsetPose_ is the inverse transformation of targetPoseW_ * prevTargetPoseW_.inv()
+    holdOffsetPose_ = prevTargetPoseW_ * targetPoseW_.inv();
+    // Transform to target pose frame (see compliancePose implementation)
+    sva::PTransformd T_0_d(targetPoseW_.rotation());
+    deltaCompPoseW_ = deltaCompPoseW_ * T_0_d.inv() * holdOffsetPose_ * T_0_d;
+  }
+  else
+  {
+    holdOffsetPose_ = sva::PTransformd::Identity();
+  }
+  prevTargetPoseW_ = targetPoseW_;
```

Comment:
I think I understand the problem you are facing but I don't think I understand what the implementation of `hold` is trying to do

1. Passing this modified `deltaCompPoseW_` to `target()` is similar to using the un-modified `deltaCompPoseW_` and `prevTargetPoseW_` in `target()`
2. But, maybe more importantly, this modification is also similar to having `deltaCompPoseW_` follow the changes in `targetPoseW_` (but with the compliance equation still kicking in) as is seen in the second graph with hold enabled

Maybe (1) is not so important for the hold effect and (2) is where all the magic is? Maybe this could be better reflected in the implementation comment?

I think it's still moving a bit because doing so reduces the force error and it becomes more acceptable as the position error becomes smaller (as the force smoothly reaches zero in your second graph with hold enabled), since you are targetting a force of zero but not quite reaching it after 10 seconds.

Reply:
I updated the document of `hold` method (https://github.com/jrl-umi3218/mc_rtc/pull/143/files#diff-db3809a3a52f623c53b2c5836c7546b76608fe22dd5e5e097bfc2c1e306e551dR205) and added a comment to refer to it for details.